File: src/literal.ml
type t

external int32 : int32 -> t = "caml_binaryen_literal_int32"
external int64 : int64 -> t = "caml_binaryen_literal_int64"
external float32_bits : int32 -> t = "caml_binaryen_literal_float32_bits"
external float64_bits : int64 -> t = "caml_binaryen_literal_float64_bits"

let float32 n = float32_bits @@ Int32.bits_of_float n

external float64 : float -> t = "caml_binaryen_literal_float64"
external vec128 : int array -> t = "caml_binaryen_literal_vec128"

let vec128 (low, low_mid, high_mid, high) =
  let bytes = Bytes.create 16 in
  Bytes.set_int32_le bytes 0 low;
  Bytes.set_int32_le bytes 4 low_mid;
  Bytes.set_int32_le bytes 8 high_mid;
  Bytes.set_int32_le bytes 12 high;
  let arr = Array.init 16 (Bytes.get_uint8 bytes) in
  vec128 arr

File: src/literal.js
//Provides: caml_binaryen_literal_int32
function caml_binaryen_literal_int32(i32) {
  // Hack around Binaryen.js
  return { type: 'int32', value: i32 };
}

//Provides: caml_binaryen_literal_int64
function caml_binaryen_literal_int64(i64) {
  // Hack around Binaryen.js
  return { type: 'int64', value: i64 };
}

//Provides: caml_binaryen_literal_float64
function caml_binaryen_literal_float64(f64) {
  // Hack around Binaryen.js
  return { type: 'float64', value: f64 };
}

//Provides: caml_binaryen_literal_float32_bits
function caml_binaryen_literal_float32_bits(i32) {
  // Hack around Binaryen.js
  return { type: 'float32_bits', value: i32 };
}

//Provides: caml_binaryen_literal_float64_bits
function caml_binaryen_literal_float64_bits(i64) {
  // Hack around Binaryen.js
  return { type: 'float64_bits', value: i64 };
}

//Provides: caml_binaryen_literal_vec128
//Requires: caml_js_from_array
function caml_binaryen_literal_vec128(v128) {
  // Hack around Binaryen.js
  return { type: "vec128", value: caml_js_from_array(v128) };
}
